Remove cases of AT_FORALL_SCALAR_TYPES_EXCEPT_HALF.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23340 [WIP] Remove unused parameter from FORALL macros and rename STUBS to QINTS.
* #23339 [WIP] Align AT_FORALL macros with AT_DISPATCH macros.
* #23338 [WIP] Don't redefine unecessary type stub.
* #23337 [WIP] Support torch::tensor and at::tensor with bool and BFloat16 dtypes.
* #23336 [WIP] Rename AT_FORALL_SCALAR_TYPES_WITH_COMPLEX to AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_STUBS
* #22932 Remove AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF, which isn't used anymore.
* #22931 Dont allow conversions with QInt.
* **#22890 Remove cases of AT_FORALL_SCALAR_TYPES_EXCEPT_HALF.**
* #22889 C++ at::Tensor, torch::tensor constructors should not accept QInts.

It's only used in some JIT printing, so doesn't necessitate its own macro definition.

Differential Revision: [D16467980](https://our.internmc.facebook.com/intern/diff/D16467980)

gchanan committed:
Remove cases of AT_FORALL_SCALAR_TYPES_EXCEPT_HALF.

It's only used in some JIT printing, so doesn't necessitate its own macro definition.

gchanan commented:
@mruberry it looks like you wrote the code (in codegen.cpp) in question.  Do you remember what this is trying to accomplish?  Is it to get a usable C++ type?  A debuggable string that is similar to the torch dtype?

mruberry commented:
@gchanan I can't take credit since I just moved/refactored the code in question, but I think I can explain it. It is to get a usable C++ type. That is, it's used to generate the type in lines of code like:

`float f = tensor[idx];

In the above, the string "float" would be generated by this function. 

Halfs are special-cased to return the string "half" instead of "at::half" because the CUDA compiler can't understand the latter. Half being special-cased *should* also make the removal of the original macro safe. 

Is that the explanation you were looking for?

gchanan commented:
@mruberry thanks, that makes sense.  This was messed up before, and is still messed up (because e.g. BFloat and QInt won't work here), but I think this change is strictly a net-positive now.

gchanan committed:
Update on "Remove cases of AT_FORALL_SCALAR_TYPES_EXCEPT_HALF."


It's only used in some JIT printing, so doesn't necessitate its own macro definition.

gchanan committed:
Update on "Remove cases of AT_FORALL_SCALAR_TYPES_EXCEPT_HALF."


It's only used in some JIT printing, so doesn't necessitate its own macro definition.

Differential Revision: [D16467980](https://our.internmc.facebook.com/intern/diff/D16467980)
